### PR TITLE
ADDS SPRESS

### PR DESCRIPTION
**Pull request by becw**
To test:
* Follow the instructions in the README that describe how to use this: https://github.com/palantirnet/butler/tree/spress-task#add-butler-to-a-project-with-spress

The styleguide template itself is not complete yet; @MichellanneLi and I are working on it here: https://github.com/palantirnet/uchicago-law/tree/experimental-styleguide/styleguide

This approach is kind of different from the current butler approach because it puts the styleguide and all of its dependencies within the `styleguide` directory; it is more independent but also more isolated.

By the way, if you google spress stuff, add `-express` because google wants to expand your results: https://www.google.com/search?q=spress+-express+config

### CODE REVIEW

**Comment by MichellanneLi** [comment]
@becw I noticed that none of the components, elements, etc. are showing up on the index page. Do I need to manually add them? In the past, we've manually added templates, but this piece of code has handled components and elements:

`{% for styleguide in data.styleguides %}
      {% if styleguide.section == "component" %}
        <li><a href="{{ site.url }}{{ styleguide.url }}">{{ styleguide.title }}</a></li>
      {% endif %}
      {% endfor %}`

**Comment by MichellanneLi** [comment]
@becw I noticed that, for the layouts, there's a placeholder variable in the front matter. What's that for?

**Comment by becw** [comment]
Questions from DEFEND:

* In spress, since you can't use extends in the layouts, will we be able to reuse these for Drupal? Drupal's twig has issues with the yaml headers in Pattern Lab, and would have the same issue here.
* This should be using `.html.twig` extensions instead of `.html` extensions for twig files.

**Comment by becw** [comment]
@MichellanneLi that placeholder variable... the page wasn't outputting what I expected, so I was just trying to debug, and it wasn't really working!

File: README.md [hunk 1]
```diff
@@ -42,6 +42,40 @@ A tool to automate front-end development tasks and streamline prototyping.
...
+1. Add Butler's scripts to your `package.json` file:
+
+  ```json
+"scripts": {
+  "butler": "node_modules/butler/node_modules/.bin/gulp --gulpfile node_modules/butler/Gulpfile.js",
+  "develop": "node_modules/butler/node_modules/.bin/gulp --gulpfile node_modules/butler/Gulpfile.js develop",
+  "tests": "node_modules/butler/node_modules/.bin/gulp --gulpfile node_modules/butler/Gulpfile.js test",
+  "deploy": "node_modules/butler/node_modules/.bin/gulp --gulpfile node_modules/butler/Gulpfile.js deploy"
+},
+```
```
Comment (labbydev) [inline README.md]: The incorrect spacing/lack of indents here kind of drives me nuts....

File: README.md [hunk 2]
```diff
@@ -42,6 +42,40 @@ A tool to automate front-end development tasks and streamline prototyping.
...
+1. Copy the contents of Butler's `STYLEGUIDE_TEMPLATE_SPRESS` directory into your `styleguide` directory with the following two commands:
+
+  ```
+cp -r node_modules/butler/STYLEGUIDE_TEMPLATE_SPRESS/* .
+cp -r node_modules/butler/STYLEGUIDE_TEMPLATE_SPRESS/.gitignore .
+```
```
Comment (labbydev) [inline README.md]: @becw is there anyway to automate this step?

File: README.md [hunk 3]
```diff
@@ -42,6 +42,40 @@ A tool to automate front-end development tasks and streamline prototyping.
...
+  "develop": "node_modules/butler/node_modules/.bin/gulp --gulpfile node_modules/butler/Gulpfile.js develop",
+  "tests": "node_modules/butler/node_modules/.bin/gulp --gulpfile node_modules/butler/Gulpfile.js test",
+  "deploy": "node_modules/butler/node_modules/.bin/gulp --gulpfile node_modules/butler/Gulpfile.js deploy"
+},
+```
+1. Copy the contents of Butler's `STYLEGUIDE_TEMPLATE_SPRESS` directory into your `styleguide` directory with the following two commands:
+
+  ```
+cp -r node_modules/butler/STYLEGUIDE_TEMPLATE_SPRESS/* .
+cp -r node_modules/butler/STYLEGUIDE_TEMPLATE_SPRESS/.gitignore .
```
Comment (labbydev) [inline README.md]: @becw this returns a `No such file or directory`

File: README.md [hunk 4]
```diff
@@ -42,6 +42,40 @@ A tool to automate front-end development tasks and streamline prototyping.
...
+```
+1. Install spress: `composer install`
+1. Add your fresh styleguide code to your project: `git add .`
```
Comment (labbydev) [inline README.md]: I would recommend running butler and testing things before committing the code (moving this step down in the process)

File: README.md [hunk 5]
```diff
@@ -42,6 +42,40 @@ A tool to automate front-end development tasks and streamline prototyping.
   * `http://[project].local:8000` if you're running Butler on a Vagrant
   * [http://localhost:8000](http://localhost:8000) if you're developing locally
 
+## Add Butler to a project, with spress
+
+1. Log in to your vagrant environment and navigate to your project. You should always run `npm` commands from within your Vagrant.
```
Comment (labbydev) [inline README.md]: I would add in the commands here: `vagrant up` `vagrant ssh` `cd /var/www/[project].local`. I read this the first time and was like "yeah high fives, good node that it should be run inside vagrant" and then proceeded to _not_ run it inside vagrant. But I could also be especially neglectful.

File: README.md [hunk 6]
```diff
@@ -42,6 +42,40 @@ A tool to automate front-end development tasks and streamline prototyping.
...
+1. Create a `styleguide` directory in your project
+1. **You'll install the styleguide dependencies and run Butler from the `styleguide` directory.** Go there now: `cd styleguide`
+1. Run `npm init` to create your `package.json` file
+1. Add Butler as a dependency:
+
+  ```
+npm install --save-exact palantirnet/butler#spress-task
```
Comment (labbydev) [inline README.md]: This looks like it needs some updates to butler's default `postinstall`:
`````
config/butler.config.sh: 4: config/butler.config.sh: cannot create ../../conf/butler.defaults.js: Directory nonexistent
config/butler.config.sh: 5: config/butler.config.sh: cannot create ../../conf/butler.defaults.js: Directory nonexistent
config/butler.config.sh: 6: config/butler.config.sh: cannot create ../../conf/butler.defaults.js: Directory nonexistent
`````

**Review by labbydev — COMMENTED** [review]
I added a couple of notes (mostly documentation based). I wasn't able to get it up and running locally or via vagrant. Every time I tried I got that it couldn't connect to server. Will try again to make sure it's not user error.

File: README.md (hunk 1 above)
Comment (becw) [inline README.md]: Me too. I tried to fix it but it would not do my bidding.

File: README.md (hunk 2 above)
Comment (becw) [inline README.md]: I guess we could do it in butler's post-install thingy.

File: README.md (hunk 3 above)
Comment (becw) [inline README.md]: Gah! It looks like `npm` strips the .gitignore files from the copy of butler! wtf.

**Comment by becw** [comment]
Ok @labbydev -- I've updated the README and revamped the install script to do more of your work for you. Thanks for testing!!

**Comment by MichellanneLi** [comment]
@becw The more I use this version of butler, the more I notice that spress is much slower than sculpin. Is there anything we can do about that?

**Comment by becw** [comment]
I was wondering that too, @MichellanneLi! I don't understand why it's so slow, since it's just serving us rendered files! I will look into it.

**Comment by labbydev** [comment]
This all looks good to me. I will say I noticed it was about 10-15seconds for butler to recognize changes I made when working with this version.

**Comment by MichellanneLi** [comment]
@becw "npm run butler" is not compiling scss changes dynamically. Rather, I have to run it again in order to see my changes.

**Comment by MichellanneLi** [comment]
@becw @labbydev When running butler locally, I repeatedly get this error message:

`[12:06:06] Finished 'sass' after 1.06 s
[12:06:06] 'spress-serve' errored after 1.42 s
[12:06:06] Error in plugin 'gulp-exec'
Message:
    Command failed: ../../vendor/bin/spress site:build --server --source=../../


  [React\Socket\ConnectionException]
  Could not bind to tcp://0.0.0.0:4000: Address already in use


site:build [-s|--source SOURCE] [--timezone TIMEZONE] [--env ENV] [--watch] [--server] [--drafts] [--safe]


Details:
    killed: false
    code: 1
    signal: null
    cmd: ../../vendor/bin/spress site:build --server --source=../../`

**Comment by aCyborg** [comment]
I'm going to pile on and echo Miche's comment about the sass not compiling unless you rerun it. I thought about maybe having the sass compile into two locations (build and src) but I'm not sure how that will muck things up or if it will even work. I tried changing it in the gulpfile, but it seems it is configured to only render in one place so it won't accept multiple dirs.

**Comment by MichellanneLi** [comment]
"npm run deploy" is not working. :(

**Comment by labbydev** [comment]
we should definitely run npm-shrinkwrap on this to make sure we lock down the versions of things.

**Comment by patrickfweston** [comment]
Not sure if I should push to this PR or not, but on line 103 of the Gulpfile, adding in the CSS path ensures that spress is rebuilt after a Sass change:

```
var watcher = gulp.watch([defaults.spress_home + 'src/content/*', defaults.spress_home + 'src/**/*.html*', defaults.spress_home + 'src/content/assets/css/*'], ['spress-build']);
```

Without this, changes to Sass alone will trigger only a Sass recompile. This will not update the CSS file used by Spress because `spress-watch` is not triggered.

I'll create a PR against the spress-task branch and assign it to @aCyborg and @becw for review.

**Pull request merged**